From e1534020ffbc049495a55b35b71b63aa478cf8de Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Zolt=C3=A1n=20B=C3=B6sz=C3=B6rm=C3=A9nyi?=
 <zboszor@gmail.com>
Date: Tue, 29 Apr 2025 12:11:09 +0200
Subject: [PATCH] Fix building lldb with Python 3.13
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

This change backports some pieces from the LLVM project.

Upstream-Status: Backport
Signed-off-by: Zoltán Böszörményi <zboszor@gmail.com>
---
 .../Python/PythonDataObjects.cpp              |  6 ++----
 .../Python/ScriptInterpreterPython.cpp        | 19 +++++++++++--------
 2 files changed, 13 insertions(+), 12 deletions(-)

diff --git a/lldb/source/Plugins/ScriptInterpreter/Python/PythonDataObjects.cpp b/lldb/source/Plugins/ScriptInterpreter/Python/PythonDataObjects.cpp
index 68f4e90d70f6..6e783a0f2284 100644
--- a/lldb/source/Plugins/ScriptInterpreter/Python/PythonDataObjects.cpp
+++ b/lldb/source/Plugins/ScriptInterpreter/Python/PythonDataObjects.cpp
@@ -70,10 +70,8 @@ Expected<std::string> python::As<std::string>(Expected<PythonObject> &&obj) {
 }
 
 static bool python_is_finalizing() {
-#if PY_MAJOR_VERSION == 2
-  return false;
-#elif PY_MAJOR_VERSION == 3 && PY_MINOR_VERSION < 7
-  return _Py_Finalizing != nullptr;
+#if PY_VERSION_HEX >= 0x030d0000
+  return Py_IsFinalizing();
 #else
   return _Py_IsFinalizing();
 #endif
diff --git a/lldb/source/Plugins/ScriptInterpreter/Python/ScriptInterpreterPython.cpp b/lldb/source/Plugins/ScriptInterpreter/Python/ScriptInterpreterPython.cpp
index 1bf647e4acfc..a974f4877185 100644
--- a/lldb/source/Plugins/ScriptInterpreter/Python/ScriptInterpreterPython.cpp
+++ b/lldb/source/Plugins/ScriptInterpreter/Python/ScriptInterpreterPython.cpp
@@ -188,19 +188,22 @@ private:
   }
 
   void InitializeThreadsPrivate() {
-// Since Python 3.7 `Py_Initialize` calls `PyEval_InitThreads` inside itself,
-// so there is no way to determine whether the embedded interpreter
-// was already initialized by some external code. `PyEval_ThreadsInitialized`
-// would always return `true` and `PyGILState_Ensure/Release` flow would be
-// executed instead of unlocking GIL with `PyEval_SaveThread`. When
-// an another thread calls `PyGILState_Ensure` it would get stuck in deadlock.
-#if (PY_MAJOR_VERSION == 3 && PY_MINOR_VERSION >= 7) || (PY_MAJOR_VERSION > 3)
+    // Since Python 3.7 `Py_Initialize` calls `PyEval_InitThreads` inside itself,
+    // so there is no way to determine whether the embedded interpreter
+    // was already initialized by some external code. `PyEval_ThreadsInitialized`
+    // would always return `true` and `PyGILState_Ensure/Release` flow would be
+    // executed instead of unlocking GIL with `PyEval_SaveThread`. When
+    // an another thread calls `PyGILState_Ensure` it would get stuck in deadlock.
+
     // The only case we should go further and acquire the GIL: it is unlocked.
     if (PyGILState_Check())
       return;
-#endif
 
+#if PY_VERSION_HEX < 0x03090000
     if (PyEval_ThreadsInitialized()) {
+#else
+    if (true) {
+#endif
       Log *log(lldb_private::GetLogIfAllCategoriesSet(LIBLLDB_LOG_SCRIPT));
 
       m_was_already_initialized = true;
-- 
2.49.0

